### ISSUE TITLE
[Bug]: store and forward chronology

### ISSUE THREAD

**Issue by jranma** [meshtastic/firmware#4166]
### Category

Other

### Hardware

T-Beam S3

### Firmware Version

2.3.12

### Description

when one retrieves store and forward messages using "SF", all messages will arrive at the same time. Above all, they will all be displayed with the time they were retrieved, not the time they were received on the relay. It is therefore difficult, if not impossible, to understand the thread of the conversation because the chronologies are wrong.

**Comment by garthvh** [comment]
Where are you seeing the messages with the same time?

**Comment by jranma** [comment]
each time I request messages with a "SF" command, I get all the messages with the same time, which is the time at wich I requested them. 
Using Android APP.

**Comment by GUVWAF** [comment]
Yes, that sounds correct. Currently the timestamp when the packet was actually received is not stored and send with the packet upon request.

**Comment by jranma** [comment]
OK, not sure the messages come in the right order. 
Last time I received many messages and as the timestamp are all the same, I had a hard time understanding the logic of the conversation. 
Having the message sorted by their "real" timestamp would be a nice emprovement.

**Comment by GUVWAF** [comment]
I think we can make this work. Currently we do store a relative time in the StoreForward history, but we can change it to an absolute time if we have one, and then not updating the `rx_time` for StoreForward packets when requesting them.

**Comment by NeimadTL** [comment]
@GUVWAF @caveman99 @fifieldt , is this still up for grab ? Not familiar at all with the project but looks this [method](https://github.com/meshtastic/firmware/blob/develop/src/mesh/MeshService.cpp#L456) needs to be updated and this [line](https://github.com/meshtastic/firmware/blob/develop/src/mesh/MeshService.cpp#L195) needs to be removed based on what has been discussed already, correct ?

**Comment by GUVWAF** [comment]
@NeimadTL Yes, it's still on the TODO, but to be honest it's more involved than you might think. You would have to add the original time somewhere here: https://github.com/meshtastic/firmware/blob/develop/src/modules/StoreForwardModule.cpp#L195, and then also send that with the packet. So it needs to be added to the protobuf definition of a StoreForward message: https://github.com/meshtastic/protobufs/blob/cb1f89372a70b0d4b4f8caf05aec28de8d4a13e0/meshtastic/storeforward.proto

And then you'd likely need to add an `alterReceived()` method to the StoreForwardModule to overwrite the `rx_time` with the original time from the StoreForward message.

**Comment by NeimadTL** [comment]
@GUVWAF, thank you for your hints. Just started a draft, please let me know if this is a good start or if I'm going at the wrong. way, appreciate it.
